### PR TITLE
new handler for SUSH

### PR DESCRIPTION
**Pull request by d034973**
object handler for SUSH was created

### CODE REVIEW

**Comment by larshp** [comment]
thanks! 👍 

https://docs.abapgit.org/development/guidelines.html

abapGit targets 702 syntax so some of the statements needs downporting, plus also please pretty print with upper case keywords

**Comment by larshp** [comment]
#1582

**Comment by larshp** [comment]
https://github.com/abapGit-tests/SUSH created, please open a PR with a small SUSH example, so we have it for testing in the future

File: src/objects/zcl_abapgit_object_sush.clas.abap [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+      lv_text = 'Lead application of object &1 does not exist package &2'(002).
+      REPLACE '&1' WITH is_head-name INTO lv_text.
+      REPLACE '&2' WITH iv_package INTO lv_text.
+      lv_text1 = lv_text(50).
+      lv_text2 = lv_text+50(50).
+      MESSAGE s471(s#) WITH lv_text1 lv_text2 INTO ld_msg.
+      RAISE EXCEPTION TYPE cx_su2n_raise_events EXPORTING textid = cl_su2x=>convert_to_exception( ).
+    ENDIF.
+
+    IF cl_abapgit_factory=>get_environment( )->is_sap_cloud_platform( ). "[A4C_AGIT]
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_sush.clas.abap:70]:
```suggestion
    IF cl_abapgit_factory=>get_environment( )->is_sap_cloud_platform( ) = abap_true. "[A4C_AGIT]
```

**Comment by larshp** [comment]
due to the way abapGit is distributed, via copy-paste we cannot use text elements in the code

**Comment by mbtools** [comment]
Unfortunately, the following objects do not in 702:
- `cl_su22_adt_object` 
- `if_su22_adt_object` 
- `cx_su2n_raise_events` 
- `cl_su22_appl`
- `cl_su2x`
- `cl_suso`

These will have to be replaced with generic objects, too. 

Replace the `CATCH cx_su2n_raise_events INTO lr_err.` with:

```abap
    CATCH cx_static_check INTO lr_err
        zcx_abapgit_exception=>raise( iv_text     = lx_err->get_text( )
                                      ix_previous = lx_err ).
```

File: src/objects/zcl_abapgit_object_sush.clas.abap [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+    ASSERT NOT ms_item-obj_name IS INITIAL.
+    CREATE OBJECT lo_su22.
+
+    ls_key = ms_item-obj_name.
+
+    TRY.
+        lo_su22->if_su22_adt_object~delete( iv_key       = ls_key
+                                            iv_cleanup   = abap_true   ).
+      CATCH cx_su2n_raise_events INTO lx_msg.
```
Comment (mbtools) [inline src/objects/zcl_abapgit_object_sush.clas.abap:41]:
```abap
zcx_abapgit_exception=>raise( lx_msg->get_text( ) ).
```

**Comment by larshp** [comment]
https://github.com/abapGit-tests/SUSH created, please open a PR with a small SUSH example, so we have it for testing in the future

File: src/objects/zcl_abapgit_object_sush.clas.abap [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+  PRIVATE SECTION.
+ENDCLASS.
+
+
+
+CLASS ZCL_ABAPGIT_OBJECT_SUSH IMPLEMENTATION.
+
+
+  METHOD zif_abapgit_object~changed_by.
+    rv_user = c_user_unknown.
```
Comment (mbtools) [inline src/objects/zcl_abapgit_object_sush.clas.abap:23]:
```abap
    DATA ls_key TYPE usobkey.

    ls_key = ms_item-obj_name.

    SELECT SINGLE modifier FROM usob_sm INTO rv_user
      WHERE name = ls_key-name AND type = ls_key-type.
    IF sy-subrc <> 0.
      rv_user = c_user_unknown.
    ENDIF.
```

File: src/objects/zcl_abapgit_object_sush.clas.xml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<abapGit version="v1.0.0" serializer="LCL_OBJECT_CLAS" serializer_version="v1.0.0">
+ <asx:abap xmlns:asx="http://www.sap.com/abapxml" version="1.0">
+  <asx:values>
+   <VSEOCLASS>
+    <CLSNAME>ZCL_ABAPGIT_OBJECT_SUSH</CLSNAME>
+    <LANGU>E</LANGU>
+    <DESCRIPT>SAP-specific abapGIT object Handler for SUSH</DESCRIPT>
```
Comment (mbtools) [inline src/objects/zcl_abapgit_object_sush.clas.xml:8]: Change to `SUSH - Assignment: Service to Authorization Objects`

**Comment by mbtools** [comment]
Good work. It activates without errors on 702. Just two small changes (see above).

**Review by mbtools — CHANGES_REQUESTED** [review]

**Review by mbtools — APPROVED** [review]
Great. Thanks!

**Comment by mbtools** [comment]
Just a small example for testing, please
https://github.com/abapGit-tests/SUSH

**Comment by d034973** [comment]
> Just a small example for testing, please
> https://github.com/abapGit-tests/SUSH

I sorry I just new here, I don't know where and how I have to describe an example for testing.

**Comment by mbtools** [comment]
No problem. We will help :-) 

1. Go to https://github.com/abapGit-tests/SUSH and create a fork
2. In your system (where you have the new SUSH class), start abapGit and create a new online repo for the URL of your fork (should be https://github.com/d034973/SUSH). You can pick any local package like $SUSH. 
3. Add one SUSH object to this package (like ZAG_UNIT_TEST)
4. Go back to abapGit and you should see the new object in the list for the repo
5. Now stage everything and commit. 
6. The go to your forked repo and create a pull request (at the top there's a section showing the delta to the original and a compare button which you can click to create the PR)

**Comment by larshp** [comment]
@d034973 did the above help? 

Marc also added it to https://docs.abapgit.org/development/serializers.html#testing

**Comment by d034973** [comment]
> @d034973 did the above help?
> 
> Marc also added it to https://docs.abapgit.org/development/serializers.html#testing

Hi guys, I am still on this topic. We have internal problem with our abapGit. The responsible colleagues are informed and I am waiting for update from this team. As soon our tool works fine, I will continue to create the tests

**Comment by larshp** [comment]
I'll go ahead and merge this, feel free to open a new PR with additional changes. And add the SUSH test when possible, https://github.com/abapGit-tests/SUSH

welcome to abapGit as a contributor 🎉

**Comment by d034973** [comment]
@mbtools I have created the test for SUSH, could you please check it?

**Pull request merged**